pandas/io/orc.py

```diff
@@ -55,3 +56,74 @@ def read_orc(
     with get_handle(path, "rb", is_text=False) as handles:
         orc_file = orc.ORCFile(handles.handle)
         return orc_file.read(columns=columns, **kwargs).to_pandas()
+
+
+def to_orc(
+    df: DataFrame,
+    path: FilePathOrBuffer = None,
+    engine: str = 'pyarrow',
+    index: bool = None,
+    **kwargs
+) -> bytes:
+    """
+    Write a DataFrame to the ORC format.
+    Parameters
+    ----------
+    df : DataFrame
+    path : str or file-like object, default None
+        If a string, it will be used as Root Directory path
+        when writing a partitioned dataset. By file-like object,
+        we refer to objects with a write() method, such as a file handle
+        (e.g. via builtin open function). If path is None,
+        a bytes object is returned. Note that currently the pyarrow
+        engine doesn't work with io.BytesIO.
+    engine : {{'pyarrow'}}, default 'pyarrow'
+        Parquet library to use, or library it self, checked with 'pyarrow' name
+        and version >= 5.0.0
+    index : bool, default None
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file.
+        If ``None``, similar to ``infer`` the dataframe's index(es)
+        will be saved. However, instead of being saved as values,
+        the RangeIndex will be stored as a range in the metadata so it
+        doesn't require much space and is faster. Other indexes will
+        be included as columns in the file output.
+    kwargs
+        Additional keyword arguments passed to the engine
+    Returns
+    -------
+    bytes if no path argument is provided else None
+    """
+    if index is None:
+        index = df.index.names[0] is not None
+
+    if isinstance(engine, str):
+        engine = import_optional_dependency(engine, min_version='5.0.0')
+    else:
+        try:
+            assert engine.__name__ == 'pyarrow', "engine must be 'pyarrow' module"
+            assert hasattr(engine, 'orc'), "'pyarrow' module must have orc module"
+        except ImportError as e:
+            raise ValueError(
+                "Unable to find a usable engine; "
+                "tried using: 'pyarrow'.\n"
+                "A suitable version of "
+                "pyarrow is required for ORC support.\n"
+                "Trying to import the above resulted in these errors:"
+                f"\n - {e}"
+            )
+
+    if path is None:
+        # to bytes: pyarrow auto closes buffers hence we read a pyarrow buffer
```

Comment:
Would it be worth opening an issue at pyarrow about this? This is quite a bad behavior.

Reply:
Yes. I'm actually responsible for this in Arrow in the first place since I added the ORC writer in C++ and Python there. I will take care of this one: https://issues.apache.org/jira/browse/ARROW-14783